### ISSUE TITLE
Import DeepBlur

### ISSUE THREAD

**Issue by charlesangus** [charlesangus/DeepC#54]
Add @kingkristo's DeepBlur node from https://github.com/kingkristo/DeepBlur

**Comment by falkhofmann** [comment]
I had a brief look, build and trial. my first impression.

- it is really slow, even though it shouldn't be. Ie with 2 samples without blurring in z, it should not take multiple times longer than a regular blur. i stopped at size 10
- seems like samples getting dropped when enable "blur on z only". actually just one is left after. regardless of the incoming samples
- knob_changed callbacks were not working
- ATM its using a new DeepOutPlane rather than modifying in place. but that could be due the spatial approach
- the user defined filter vs the math below seems different
- 
----
overall it's a good idea and for sure cool to see the source code with lots of comments, but it needs a fair amount of optimization.
- i would rather then to a slim version of it and limit the features exposed to the user.

**Comment by charlesangus** [comment]
Yeah I wouldn't want it just straight copied into the repo, for sure. It's a demonstration/project node, and as you say would need a bit of a weed whacker to cut it down to the essentials.

I would see bringing Deep blurring to the project in sort of a staged approach. Step one, trim this node down to size and build it but don't expose it (like DeepCBlink). Step two,  iterate for sorting out algorithms/performance/how exactly it works. Step three, expose the node once the issues have been hashed out.

**Comment by kingkristo** [comment]
> 

These are all fair comments. This project was my introduction to Nuke, and deep compositing as a whole so there are definitely refinements to be made. I was learning how to create plugins for Nuke as I went so I'm certain much of my implementation is sub optimal, speed was a big issue I wasn't able to fix in time.

As for the abundance of features of differing usefulness, the project aimed to offer lots of different potential features and then garner feedback on them after the fact. There are definitely a few which can be cut out for inclusion here.

I appreciate the extra work you're doing @charlesangus, hopefully my project can act as a jumping off point for a more refined and optimised DeepBlur, I'll keep an eye on the changes to see where my janky solution could be improved.

**Comment by charlesangus** [comment]
Hey, @kingkristo! Thanks for jumping in. Totally understand where the node is at and why it is the way it is :)

I actually haven't gotten around to trying the node in detail, so having the different options is good to see what works and what doesn't.

And yeah, exactly, hoping to use this as a jumping off point for a production ready version.

**Comment by kingkristo** [comment]
I'm not sure how obvious all of the functionality is so if you need any
more details this is my full dissertation:
https://drive.google.com/file/d/1NDkaISHX_Zo9Zz4jRVVC5MjvndCFv6pC/view?usp=sharing

Around page 53 onwards talks about the deep blur implementation.

**Comment by CMG99** [comment]
Hi, i am following on from the work regarding DeepBlur and am implementing a DeepBlur and DeepDefocus node (for university) and was wondering if you may want to incorporate it into DeepC ?

**Comment by charlesangus** [comment]
I would love to! I've added you as a contributor. Please keep your work in a new branch until it's ready for prime time :)

There's also the work that's been done by kingkristo above which may be interesting as reference or a jumping off point.

**Comment by CMG99** [comment]
first version of a DeepCBlur node on my branch `CMG-blurring-dev`

Only been testing with 1 deep image, as I don't have access to anymore

been developing on windows, and new to cmake, so may need some help getting linux builds

2D gaussian blurs should work, 3D blur might crash as I have no sample count optimizations as the final images are very large

**Comment by charlesangus** [comment]
Cool!

There's some example deep images in the example repo which could be useful,
you can also make your own examples with the ScanlineRender node,
checkerboards, and primitives

For Linux build, should be fairly straightforward. I recommend getting a
CentOS VM set up with the appropriate devtools package for the vfx
reference platform of the nuke you're targeting.

**Comment by CMG99** [comment]
The Nuke documentation references devtools-2 with GCC 6.3.1, and the docs in this repo reference devtoolset-3, however I am only able to download devtoolset-7. As a result the DeepC plugins crash when added into Nuke, so do you have any advice as to how I can continue, as I cannot seem to find any older devtoolsets or GCC 6.3.1

Thanks,
Callum

**Comment by charlesangus** [comment]
Yeah, that's gotten a bit messy... It should work with the relevant VFX Platform version (CY2019 for Nuke 12) which is devtoolset-6 if I'm not mistaken.

These may be helpful?

https://vfxplatform.com/#footnote-gcc6

https://groups.google.com/g/vfx-platform-discuss/c/_-_CPw1fD3c

https://github.com/charlesangus/DeepC/issues/53

And if you get it sorted, you would be a saint to update the build guide (which as you have noticed is cast in amber lol).

**Comment by CMG99** [comment]
I might not be be able to get a CentOS build for the plugins, as my use of templates while developing with MSVC seems to be treated differently when compiling with gcc. (dependent names and lookup issues)

And I don't understand how to solve my hundreds of errors relating to this.

**Comment by CMG99** [comment]
Can I create a new release with Windows version of my plugins as I need a single place distribute them for a user study ?

**Comment by charlesangus** [comment]
Yeah I think that's fine, go for it. Just make clear in the release notes
that the previous version is there for Linux support.

**Comment by CMG99** [comment]
Finally got around to fixing all the linux build issues. Plugins fot the latest 2 Nuke versions should be added to my prototype release from eairler.

Hopefully the plugins work and if not just let me know.
